File: leetcode/64_minimum_path_sum.py
class Solution:

    def minPathSum(self, grid):
        if not any(grid):
            return 0
        m = len(grid)
        n = len(grid[0])
        for i in range(1, m):
            grid[i][0] += grid[i-1][0]
        for j in range(1, n):
            grid[0][j] += grid[0][j-1]
        for y in range(1, m):
            for x in range(1, n):
                top = grid[y-1][x]
                left = grid[y][x-1]
                grid[y][x] += min(top, left)
        return grid[-1][-1]

File: leetcode/java/64MinimumPathSum.java
/**
 * T: O(R*C) S: O(1)
 *
 * <p>The minimum sum to reach each grid cell is determinted by the minimum sum to reach either cell
 * to the top or to the left of it. This is true only in this specific case because all cells are
 * guaranteed to have non-negative values. Due to that, we can greedily calculate the sum of the
 * path to the specific cell.
 */
class Solution {
  public int minPathSum(int[][] grid) {
    if (grid.length == 0) {
      return 0;
    }
    int rowCount = grid.length;
    int colCount = grid[0].length;
    for (int r = 0; r < rowCount; r++) {
      for (int c = 0; c < colCount; c++) {
        if (r == 0 && c == 0) {
          continue;
        }
        int top = r == 0 ? Integer.MAX_VALUE : grid[r - 1][c];
        int left = c == 0 ? Integer.MAX_VALUE : grid[r][c - 1];
        grid[r][c] += Math.min(top, left);
      }
    }
    return grid[rowCount - 1][colCount - 1];
  }
}
